GaP internal geometry

LeireLarizgoitia opened the pull request:
This is the full geometry of the internal part of the GaP detector

Commit:
Path modified in adelma.conf

Commit:
Geometry updates based on new STEP

Dimensions of vessel have been changed according to the new STEP
file. Gas volume modified so that it is centered. Corresponding
materials have been added (quartz and tpb with optical properties).

Only GasDrift and GasEL left are as Gas region generators.
Buffer deleted. Relocation of cathode and its bracket.
4 Peek Mesh holders have been added 'E' shape. Steel cylinder bar
added, joins the mesh holders with the PMT clad
4 Peek Anode holder have been added 'downside L' shape.

Quartz window with evaporated tpb (3micrometer) has been added. 4 peek
holders added 'Z' shape.

A cylindrical steel enclosure has been created to protect pmt from
high pressures. There is vacuum inside this enclosure. Corresponding
holes have been made so that the PMT do not overlap. A steel PMT
holder is added inside the vacuum cylinder. Some steel plates have
been added around the enclosure to attach it to the vessel or to
attach peek holders to this volume.

Commit:
Colors have also been assigned to volumes depending on the materials

Dark gray - Steel
Dark blue - Vacuum
Cyan - Mesh
Light gray - Mesh Brackets
White - Peek holders
Yellow - Quartz Window
Orange - TPB coating
White and blue - PMT

Aretno commented on recoGanESS/gres/config/adelma.conf:
```diff
@@ -1,8 +1,8 @@
-files_in          = '/home/ander/Work/GanESS/data/GaP_test_1000.h5'
-file_out          = '/home/ander/Work/GanESS/data/GaP_test_1000_adelma.h5'
+files_in          = '/Users/llarizgoitialive/Documents/FISIKA_PhD/GanESS/GanESS_software/GanESS/simGanESS/GaP_test.gap.h5'
+file_out          = '/Users/llarizgoitialive/Documents/FISIKA_PhD/GanESS/GanESS_software/GanESS/data/GaP_test_adelma.h5'
```
This is a file from the reco software, not the simulation, so just limit the changes in the PR to the simulation software changes.

Aretno commented on simGanESS/source/geometries/GaP.hh:
```diff
@@ -26,26 +26,57 @@ class GaP : public GeometryBase
         void DefineConfigurationParameters();
         void Construct();
-        void BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic_vessel);
+        void BuildTPC(G4Material* gas, G4Material* mesh_mat, G4Material* steel, G4Material* peek, G4Material* vacuum, G4Material* quartz, G4Material* tpb, G4LogicalVolume* logic_vessel);
```
I'm not very fond of having the materials passed as an argument here, at least as long as everything is build in one function. Still, you can keep it as is, in case you want to build future detector parts with a function outside of BuildTPC.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -287,18 +320,74 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+
+    G4Box *solid_meshHolder_block = new G4Box("MeshHolder", meshHolder_width_/2, meshHolder_thickn_/2 , meshHolder_length_/2);
+    G4Box *solid_meshHolder_hole = new G4Box("MeshHolderHole", meshHolder_width_, meshHolder_hole_thickn_ , meshHolder_hole_length_/2);
+    G4Box *solid_meshHolder_holeUp = new G4Box("MeshHolderHole", meshHolder_width_, meshHolder_hole_thickn_ , meshHolder_holeUp_length_/2);
```
Same naming ("MeshHolderHole") as the above G4Box, put a different one.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -287,18 +320,74 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    G4Box *solid_meshHolder_hole = new G4Box("MeshHolderHole", meshHolder_width_, meshHolder_hole_thickn_ , meshHolder_hole_length_/2);
+    G4Box *solid_meshHolder_holeUp = new G4Box("MeshHolderHole", meshHolder_width_, meshHolder_hole_thickn_ , meshHolder_holeUp_length_/2);
+    G4VSolid        *solidSub_meshHolder0 = new G4SubtractionSolid("MeshHolderHole", solid_meshHolder_block, solid_meshHolder_hole, 0, G4ThreeVector(0., meshHolder_thickn_/2, meshHolder_length_/2-meshHolder_hole_length_/2-5*mm) );
+    G4VSolid        *solidSub_meshHolder = new G4SubtractionSolid("MeshHolderHole", solidSub_meshHolder0, solid_meshHolder_holeUp, 0, G4ThreeVector(0., meshHolder_thickn_/2, -(meshHolder_length_/2-meshHolder_holeUp_length_/2-5*mm)) );
```
As before, try using different naming for different VSolids.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -287,18 +320,74 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    new G4PVPlacement(Rot45,   G4ThreeVector(-meshHolder_x, -meshHolder_y, -meshHolder_z), logic_meshHolder, "MeshHolder", logic_vessel, false, 0, true);
+    new G4PVPlacement(Rot135,  G4ThreeVector(-meshHolder_x, meshHolder_y, -meshHolder_z), logic_meshHolder, "MeshHolder", logic_vessel, false, 1, true);
+    new G4PVPlacement(Rot_45,  G4ThreeVector(meshHolder_x, -meshHolder_y, -meshHolder_z), logic_meshHolder, "MeshHolder", logic_vessel, false, 2, true);
+    new G4PVPlacement(Rot_135, G4ThreeVector(meshHolder_x, meshHolder_y, -meshHolder_z), logic_meshHolder, "MeshHolder", logic_vessel, false, 3, true);
```
If parts are different, use different names. I'm fine with A, B, C etc. Or other name scheme that you like but make sure you differentiate them.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -287,18 +320,74 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    new G4PVPlacement(0, G4ThreeVector(meshHolderBar_xy, meshHolderBar_xy, -meshHolderBar_z), logic_meshHolderBar, "MeshHolderBar", logic_vessel, false, 0, true);
+    new G4PVPlacement(0, G4ThreeVector(-meshHolderBar_xy, meshHolderBar_xy, -meshHolderBar_z), logic_meshHolderBar, "MeshHolderBar", logic_vessel, false, 1, true);
+    new G4PVPlacement(0, G4ThreeVector(meshHolderBar_xy, -meshHolderBar_xy, -meshHolderBar_z), logic_meshHolderBar, "MeshHolderBar", logic_vessel, false, 2, true);
+    new G4PVPlacement(0, G4ThreeVector(-meshHolderBar_xy, -meshHolderBar_xy, -meshHolderBar_z), logic_meshHolderBar, "MeshHolderBar", logic_vessel, false, 3, true);
```
As before, distinguish the part names.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -314,12 +403,40 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    new G4PVPlacement(Rot45,   G4ThreeVector(0., 0., -anodeHolder_z), logic_anodeHolder, "AnodeHolder", logic_vessel, false, 0, true);
+    new G4PVPlacement(Rot135,  G4ThreeVector(0., .0, -anodeHolder_z), logic_anodeHolder, "AnodeHolder", logic_vessel, false, 1, true);
+    new G4PVPlacement(Rot_45,  G4ThreeVector(0., 0., -anodeHolder_z), logic_anodeHolder, "AnodeHolder", logic_vessel, false, 2, true);
+    new G4PVPlacement(Rot_135, G4ThreeVector(0., 0., -anodeHolder_z), logic_anodeHolder, "AnodeHolder", logic_vessel, false, 3, true);
```
Part names again.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +449,173 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    new G4PVPlacement(Rot45,   G4ThreeVector(0., 0., -quartz_windowHolder_z), logic_windowHolder, "QuartzWindowHolder", logic_vessel, false, 0, true);
+    new G4PVPlacement(Rot135,  G4ThreeVector(0., 0., -quartz_windowHolder_z), logic_windowHolder, "QuartzWindowHolder", logic_vessel, false, 1, true);
+    new G4PVPlacement(Rot_45,  G4ThreeVector(0., 0., -quartz_windowHolder_z), logic_windowHolder, "QuartzWindowHolder", logic_vessel, false, 2, true);
+    new G4PVPlacement(Rot_135, G4ThreeVector(0., 0., -quartz_windowHolder_z), logic_windowHolder, "QuartzWindowHolder", logic_vessel, false, 3, true);
```
Naming

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +449,173 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    new G4PVPlacement(Rot45,   G4ThreeVector(0., 0., -quartz_windowHolderTop_z), logic_windowHolderTop, "QuartzWindowHolderTop", logic_vessel, false, 0, true);
+    new G4PVPlacement(Rot135,  G4ThreeVector(0., 0., -quartz_windowHolderTop_z), logic_windowHolderTop, "QuartzWindowHolderTop", logic_vessel, false, 1, true);
+    new G4PVPlacement(Rot_45,  G4ThreeVector(0., 0., -quartz_windowHolderTop_z), logic_windowHolderTop, "QuartzWindowHolderTop", logic_vessel, false, 2, true);
+    new G4PVPlacement(Rot_135, G4ThreeVector(0., 0., -quartz_windowHolderTop_z), logic_windowHolderTop, "QuartzWindowHolderTop", logic_vessel, false, 3, true);
```
Naming

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +449,173 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    G4double enclosure_pmt_z = vessel_length_/2 - enclosure_pmt_length_/2;
+    G4double relative_pmt_z  = enclosure_pmt_z - pmt_z;
+
+    G4Tubs *solid_pmt = new G4Tubs("SolidPMT", 0., 25.4*mm/2, 43.*mm/2, 0., 360.*deg); // Hamamatsu pmt length: 43*mm | STEP pmt gap length: 57.5*mm
```
Declare radius and thickness as variables outside the G4Tubs declaration, for reading clarity.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +449,173 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
-
+    G4double pmt_z  = 42.495*mm + pmt_length_/2;
+
+    G4ThreeVector pmt0_Ps = G4ThreeVector(-15.573*mm, -32.871*mm, -pmt_z);
```
Probably better to have an array object of XY PMT positions, and use them here and afterwards. This would allow to change the numbers easily without having to check all places where the position is used. I'd do two arrays, one for X positions, another for Y positions.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +449,173 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+
+    G4LogicalVolume *logic_enclosurevac_pmt = new G4LogicalVolume(solidSub_enclosurevac_pmt6, vacuum, "EnclosureVacPMT");
+
+    new G4PVPlacement(0, G4ThreeVector(0., 0, enclosure_pmt_z - enclosurevac_pmt_z), logic_enclosurevac_pmt, "EnclosureVacPMT", logic_enclosure_pmt, false, 0, true);
```
Are you sure about the Z position here? If this were a daughter of logic_vessel it would make sense I think, but its mother is the logic_enclosure_pmt. I'm not saying is wrong, but using the global position of the mother to position a daughter inside the mother looks suspicious (it may work but just because the centering is as is).

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +449,173 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    G4LogicalVolume *logic_plate1_pmt = new G4LogicalVolume(solidSub_plate1_pmt6, steel, "PMTplateBottom");
+
+    G4double plate1_pmt_z = enclosure_pmt_z - enclosure_pmt_length_/2 - plate_pmt_length_/2;
+    new G4PVPlacement(0, G4ThreeVector(0., 0., -plate1_pmt_z), logic_plate1_pmt, "PMTplateBottom", logic_vessel, false, 0, true);
```
In general all these lines (from the declaration of the PMTs up to here) seem non optimal. I think this should be more clean and would make more sense in a for loop (1 loop per PMT). See if you can modify the code to do that, otherwise it's too crowded.

Aretno replied:
Well, maybe up to a bit later, you'll see once you do it.

Aretno reviewed (changes requested):
Nice work! Some quick things to fix:

- Remove the change of a reco software file.
- Don't use name duplicates.
- Refactor the PMT part of the geometry using a for loop.

Commit:
Restored recoGanESS file.

Commit:
GaP internal geometry PR3 changes.

Different naming for Solids, Physical Volumes...
Radius and length of PMT as variables.
Positions of PMT as vectors. X and Y coordinates separated. Position
is defined with the corresponfing X,Y pair.
Checked that Z position is correct.

Every volume with hole for the PMTs now grouped in a for loop (1 per
PMT).

Commit:
Relocation of the Fe55 source

As the geometry has been displaced, we had to relocate de Fe55
source. Now it is just under the catode, outside the gas drift region.

Aretno commented on simGanESS/source/geometries/GaP.cc:
```diff
@@ -332,49 +453,144 @@ void GaP::BuildTPC(G4Material* gas, G4Material* mesh_mat, G4LogicalVolume* logic
+    G4double plate1_pmt_z = enclosure_pmt_z - enclosure_pmt_length_/2 - plate_pmt_length_/2;
+    G4VSolid *solid_plate1_pmt = new G4Tubs("PMTplateBottom1", 0, plate_pmt_rad_+plate_pmt_thickn_, plate_pmt_length_/2, 0, 360*deg);
+
+    G4int nHole = 7;
```
Better to use the size of the pmt_PsX vector.

Aretno reviewed (changes requested):
Just change the declaration of nHoles and use the number of elements in the PMT position vector and it'll be good to go!

Commit:
Erased declaration of nHole. Size of the position vector is used now

Aretno reviewed (approved):
Good to go!
